lua/starfall/libs_cl/websocket.lua

```diff
@@ -0,0 +1,187 @@
+	local html = assert( self.state == READYSTATE.OPEN and self.html, "WebSocket not connected" )
+	html:RunJavascript([[
+		if (sf_websocket.readyState == 1) {
+			sf_websocket.send("]] .. string.JavascriptSafe(msg) .. [[");
+		}
+	]])
+end
+
+--[[
+	End of Websocket library
+]]
+
+local checkluatype = SF.CheckLuaType
+
+--- Websocket Type. Create a websocket with websocket.new
+-- @name WebSocket
+-- @class type
+-- @libtbl websocket_methods
+-- @libtbl websocket_meta
+SF.RegisterType("WebSocket", true, false, WebSocket)
+
+return function(instance)
+
+if LocalPlayer() ~= instance.player then return end
+
+local websocket_methods, websocket_meta, wrap, unwrap = instance.Types.WebSocket.Methods, instance.Types.WebSocket, instance.Types.WebSocket.Wrap, instance.Types.WebSocket.Unwrap
+
+local websocket_list = {}
+
+--- Creates a new websocket object.
+-- @name builtins_library.WebSocket
+-- @param string addr Address of the websocket server.
+-- @param number? port Port of the websocket server. (Default 443)
+-- @param boolean? secure Whether to use secure connection (wss). (Default false)
+-- @return WebSocket The websocket object. Use WebSocket:connect() to connect.
+function instance.env.WebSocket(addr, port, secure)
+	checkluatype(addr, TYPE_STRING)
+	if port ~= nil then checkluatype(port, TYPE_NUMBER) end
+	if secure ~= nil then checkluatype(secure, TYPE_BOOL) end
+
+	local websocket =  WebSocket.new(addr, port, secure)
+	websocket_list[websocket] = true
+	return wrap(websocket)
+end
+
+--- Closes the websocket connection. Does nothing if already closed
+function websocket_methods:close()
+	unwrap(self):close()
+end
+
+--- Sends a message to the connected websocket stream.
+-- @param string msg What to send
+function websocket_methods:write(msg)
+	unwrap(self):write(msg)
+end
+
+--- Connects to the websocket server.
+function websocket_methods:connect()
+	unwrap(self):connect()
+end
+
+--- Returns the current state of the websocket.
+-- https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/readyState
+-- * 0 - CONNECTING
+-- * 1 - OPEN
+-- * 2 - CLOSING
+-- * 3 - CLOSED
+-- @return number The current state of the websocket.
+function websocket_methods:getState()
+	return unwrap(self).state
+end
+
+local Callbacks = {
+	["onMessage"] = true,
+	["onConnected"] = true,
+	["onDisconnected"] = true
+}
+
+--- Sets a callback for the websocket.
+-- Can be used with the following callbacks:
+-- * onMessage - Called when a message is received.
+-- * onConnected - Called when the websocket initially connects.
+-- * onDisconnected - Called when the websocket is disconnected, with the only param being if it was caused by an 'error' event.
+-- @param k string onMessage, onConnected, onDisconnected
+-- @param v function
+function websocket_meta:__newindex(k, v)
+	local cb = Callbacks[k]
+	if cb then
+		if type(v) == "function" then
+			unwrap(self)[k] = function(_, arg) v(self, arg) end
+		elseif v == nil then
+			unwrap(self)[k] = nil
+		end
+	end
```

Comment:
probably want a rawset otherwise, in-case someone wants to store a var on the websocket table for whatever reason.

Reply:
Still need a rawset for the else case of the `if cb then`

Reply:
Ah wasn't sure what you meant